src/core/metrics/exporter/retry_socket.rs

```diff
@@ -0,0 +1,118 @@
+//! A TCP Socket wrapper that reconnects automatically.
+use crate::prelude::{IOError, IOErrorKind, IOResult};
+use std::fmt;
+use std::net::SocketAddr;
+use std::time::Duration;
+use tokio::io::AsyncWriteExt;
+use tokio::net::TcpStream;
+use tokio::sync::oneshot::{channel, error::TryRecvError, Receiver};
+use tokio::time::interval;
+
+const MIN_RECONNECT_DELAY_MS: u64 = 50;
+const MAX_RECONNECT_DELAY_MS: u64 = 10_000;
+
+#[derive(Debug)]
+enum State {
+    Socket(TcpStream),
+    Task(Receiver<TcpStream>),
+}
+
+pub(super) struct RetrySocket {
+    address: SocketAddr,
+    socket: State,
+}
+
+impl fmt::Debug for RetrySocket {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.socket.fmt(f)
+    }
+}
+
+impl RetrySocket {
+    pub(super) async fn new(address: SocketAddr) -> Self {
+        let receiver = Self::spawn_task(address.clone());
+        RetrySocket {
+            address,
+            socket: State::Task(receiver),
+        }
+    }
+
+    fn spawn_task(address: SocketAddr) -> Receiver<TcpStream> {
+        let (tx, rx) = channel();
+        tokio::spawn(async move {
+            let mut delay = MIN_RECONNECT_DELAY_MS;
+            loop {
+                if let Ok(stream) = TcpStream::connect(address).await {
+                    if let Err(e) = tx.send(stream) {
+                        warn!("Failed to send stream from task: {:?}", e);
+                    }
+                    return;
+                }
+
+                delay = MAX_RECONNECT_DELAY_MS.min(delay << 1);
+                let mut conn_interval = interval(Duration::from_millis(delay));
```

Comment:
And for the future: you should create interval before loop and call tick on it inside.

Reply:
This interval is dynamic: it grows from min time (for example its value is 5) to max time (for example: 1000).
On first try (to connect) interval would be 5, on second - 10, on third - 20 and etc. On the ninth try this interval will become stable and equal to max time (1000). So I can't move it outside of the loop with this condition (But I would able to do this, if I make this interval constant)